src/online2bin/online2-tcp-nnet3-decode-faster.cc

```diff
@@ -0,0 +1,462 @@
+
+    TcpServer server(read_timeout);
+
+    server.Listen(port_num);
+
+    while (true) {
+
+      server.Accept();
+
+      int32 samp_count = 0;//this is used for output refresh rate
+      size_t chunk_len = static_cast<size_t>(chunk_length_secs * samp_freq);
+
+      int32 check_update = static_cast<int32>(samp_freq * output_freq);
+      int32 next_check = check_update;
+
+      int32 frame_offset = 0;
+
+      bool eos = false;
+
+      OnlineIvectorExtractorAdaptationState adaptation_state(
+          feature_info.ivector_extractor_info);
+
+      OnlineNnet2FeaturePipeline feature_pipeline(feature_info);
+      SingleUtteranceNnet3Decoder decoder(decoder_opts, trans_model,
+                                          decodable_info,
+                                          *decode_fst, &feature_pipeline);
+
+      while (!eos) {
+
+        decoder.InitDecoding(frame_offset);
+
+//        if (!adapt_speaker)
+//          feature_pipeline.SetAdaptationState(adaptation_state); //reset adaptation state to intital one
+
+        OnlineSilenceWeighting silence_weighting(
+            trans_model,
+            feature_info.silence_weighting_config,
+            decodable_opts.frame_subsampling_factor);
+
+        std::vector<std::pair<int32, BaseFloat>> delta_weights;
+
+        int32 samp_pipeline = 0;//this is used to figure out the offset for the remainder
+
+        while (true) {
+
+          eos = !server.ReadChunk(chunk_len);
+
+          if (!eos) {
+            Vector<BaseFloat> wave_part = server.GetChunk();
+            feature_pipeline.AcceptWaveform(samp_freq, wave_part);
+            samp_pipeline += wave_part.Dim();
+            samp_count += chunk_len;
+
+            if (silence_weighting.Active() &&
+                feature_pipeline.IvectorFeature() != NULL) {
+              feature_pipeline.UpdateFrameWeights(delta_weights,
+                                                  frame_offset * decodable_opts.frame_subsampling_factor);
+            }
+
+            decoder.AdvanceDecoding();
+
+            if (samp_count > next_check) {
+              if (decoder.NumFramesDecoded() > 0) {
+                Lattice lat;
+                decoder.GetBestPath(false, &lat);
+
+                std::string msg = LatticeToString(lat, word_syms);
+
+                server.Write(msg + '\r');
+              }
+
+              next_check += check_update;
+            }
+
+          } else {
+            feature_pipeline.InputFinished();
+            decoder.AdvanceDecoding();
+
+            decoder.FinalizeDecoding();
+            frame_offset += decoder.NumFramesDecoded();
+
+            if (decoder.NumFramesDecoded() > 0) {
+              CompactLattice lat;
+              decoder.GetLattice(true, &lat);
+
+              std::string msg = LatticeToString(lat, word_syms);
+
+              server.Write(msg + "\n");
+            } else
+              server.Write("\n");
+
+            server.Disconnect();
+            break;
```

Comment:
Seems like this break statement is not taking us out of the second loop. I mean we want to go to line 206 here.

Reply:
It's not a big deal, because, the next time this loop is entered the read will fail and the program does exit gracefully.

But I will fix it by setting `eos=true` after server disconnect, which will exit the outer loop as well.